msdk-io/msdk-io-mzml2/src/main/java/io/github/msdk/io/mzml2/MzMLSpectrum.java

```diff
@@ -154,95 +148,153 @@ public Integer getNumberOfDataPoints() {
 
       compressions.add(getIntensityBinaryDataInfo().getCompressionType());
 
-      InputStream encodedIs = new ByteBufferInputStreamAdapter(mappedByteBufferInputStream, 0,
+      InputStream encodedIs = new ByteBufferInputStreamAdapter(mappedByteBufferInputStream,
+          getIntensityBinaryDataInfo().getPosition(),
           getIntensityBinaryDataInfo().getEncodedLength());
       InputStream decodedIs = Base64.getDecoder().wrap(encodedIs);
       byte[] decodedData = IOUtils.toByteArray(decodedIs);
+      System.out.println(new String(IOUtils.toByteArray(encodedIs)));
 
-      result = MzMLIntensityPeaksDecoder.decode(decodedData, 0, precision,
+      intensityValues = MzMLIntensityPeaksDecoder.decode(decodedData, decodedData.length, precision,
           getIntensityBinaryDataInfo().getArrayLength(), compressions).arr;
     } catch (Exception e) {
       throw (new MSDKRuntimeException(e));
     }
 
-    return result;
+    return intensityValues;
   }
 
   @Override
-  public Float getTIC() {
-    // TODO Auto-generated method stub
-    return null;
+  public MsSpectrumType getSpectrumType() {
+    MsSpectrumType spectrumType = SpectrumTypeDetectionAlgorithm.detectSpectrumType(mzValues,
+        intensityValues, getMzBinaryDataInfo().getArrayLength());
+    return spectrumType;
   }
 
   @Override
-  public Range<Double> getMzRange() {
-    // TODO Auto-generated method stub
-    return null;
+  public Float getTIC() {
+    Float tic = MsSpectrumUtil.getTIC(intensityValues, getMzBinaryDataInfo().getArrayLength());
```

Comment:
Calculating the TIC requires loading the datapoints, so you can keep the TIC as a Float variable initially set to null, and calculate it the first time when this function is called. Then next time it does not need to be calculated again.